Create published_state field on script and unit_group

dmcavoy opened the pull request:
Create published_state field on script and unit_group

## Links

Plan: https://docs.google.com/document/d/1vUEuSp4RlldFurDYfd7sVbk1nfr3RvupimN6CzwxkZQ/edit#
Jira: https://codedotorg.atlassian.net/browse/PLAT-1059

## Testing story

None

## Follow-up work

Set published_state field based on hidden/visible and is_stable. Make sure its saved: https://github.com/code-dot-org/code-dot-org/pull/41026

Move onto using published_state in some places: https://github.com/code-dot-org/code-dot-org/pull/41027

Hamms reviewed (commented):
We should at a minimum add some Rails validation to ensure that this field is always one of the proscribed values, but I'd also be interested in considering whether or not we want to implement this as an enum rather than a string.

Also: do we want indexes, a default value, or to disallow null values?

dmcavoy commented:
> We should at a minimum add some Rails validation to ensure that this field is always one of the proscribed values, but I'd also be interested in considering whether or not we want to implement this as an enum rather than a string.
> 
> Also: do we want indexes, a default value, or to disallow null values?

Totally agree that validation, default value, and disallow null are things we should do. I think we need to wait until after we have the chance to set the value of published_state based on visible/hidden and is_stable.  I have it tracked here https://codedotorg.atlassian.net/browse/PLAT-1102 so I won't forget to come back

I don't see enum as an option for the field type so if it is I would be open to trying that. Can you point me to an example?

I'm never sure about indexes so if you think we should add one I'm all ears.

davidsbailey commented:
> I'm never sure about indexes so if you think we should add one I'm all ears.

I don't see a reason to add an index in this case, since I don't think we're going to be querying on this column. we could always add one if we start to need it.

Hamms commented:
> I don't see enum as an option for the field type so if it is I would be open to trying that. Can you point me to an example?

Ah, actually it looks like the way Rails wants you to implement enums is independent of column type: https://api.rubyonrails.org/classes/ActiveRecord/Enum.html https://naturaily.com/blog/ruby-on-rails-enum

> I don't see a reason to add an index in this case, since I don't think we're going to be querying on this column.

If we're _not_ going to be querying on this column, then there's no need to make it a column at all, and we should just throw this data in the properties hash

Hamms commented:
> Totally agree that validation, default value, and disallow null are things we should do. I think we need to wait until after we have the chance to set the value of published_state based on visible/hidden and is_stable. I have it tracked here https://codedotorg.atlassian.net/browse/PLAT-1102 so I won't forget to come back

I don't see why this would be true for anything except disallowing null; for both validation and default values, it's significantly easier to implement that functionality _before_ we have data in the column than it is to do so after.

davidsbailey commented:
> If we're _not_ going to be querying on this column, then there's no need to make it a column at all, and we should just throw this data in the properties hash

This seems like the first thing to decide. The situation is that we do not need to query this column right now, and we may or may not need to query it in the future. My understanding is that based on this information, we decided during stand up to make this a column.  Unless there is new information and a compelling reason, let's stick with the decision from stand up and keep this a column.

dmcavoy commented:
I followed the guidance at https://sipsandbits.com/2018/04/30/using-database-native-enums-with-rails/ on how to make an enum column with MySQL and ActiveRecord. I did not do the step where they talk about using db/structure.sql.

dmcavoy commented:
I ran bundle exec rake build to check that everything worked with the new validation.

davidsbailey commented:
>> I'd also be interested in considering whether or not we want to implement this as an enum rather than a string.

> I followed the guidance at https://sipsandbits.com/2018/04/30/using-database-native-enums-with-rails/ on how to make an enum column with MySQL and ActiveRecord. I did not do the step where they talk about using db/structure.sql.

Thank you Dani. I am not aware of us having a best practice on whether or not to use this pattern, but this seems like it should work fine. This will certainly ensure that this column's value will not get into a bad state, whereas rails validations are less reliable since they can sometimes get skipped. The only downside I can see is that it will take a DB migration to add a new enum value, but since we don't plan to do that often and the table is small, that should be ok.

davidsbailey commented on dashboard/db/migrate/20210609212022_add_published_state_to_script.rb:
```diff
@@ -1,5 +1,12 @@
-    add_column :scripts, :published_state, :string
+  def up
+    execute <<-SQL
+        ALTER TABLE scripts ADD published_state enum('in_development', 'pilot', 'beta', 'preview', 'stable');
```
it seems like this enum should be documented somewhere other than the migration file. was there a reason not to do the db/structure.sql step? if so, can you find a different way to document this?

dmcavoy replied:
I can add comments to make this clear. Was there another way of documenting you would think would be helpful? There is also the line in the models that starts with `enum published_state` which helps to point that that field is an enum. 

I didn't do the db/structure.sql step because they were saying to get rid of db/schema.rb and use db/structure.sql instead and I didn't think that was what we would want to do.

davidsbailey replied:
> I can add comments to make this clear. Was there another way of documenting you would think would be helpful? There is also the line in the models that starts with enum published_state which helps to point that that field is an enum.

I think the key thing to document is that if we add a new state in the future, updating the shared constants is not enough -- you'll also have to update the database to accept the new value.

> I didn't do the db/structure.sql step because they were saying to get rid of db/schema.rb and use db/structure.sql instead and I didn't think that was what we would want to do.

I see - yes, we probably don't want to make that change, then.

TBH without the ability for the enum to be reflected in schema.rb or the model file, I would vote that disadvantages of the DB enum option seem to outweigh the advantages. Let's go with the simpler option of using a string column for now, and if we decide to move in the enum direction as a team this should be easy to change later. It'll be really nice that when you update the shared constant, the validation logic will get automatically updated.

dmcavoy replied:
@Hamms any thoughts before I go back to the string? I'll wait till 5pm PT today to make the change to give you a chance to weigh in.

dmcavoy replied:
Alright its back to a string now.

Hamms replied:
string seems fine to me, but I'm glad we looked into this!

dmcavoy commented:
@Hamms @davidsbailey I think I have addressed the comments you all have made. The two things I'm holding off on right now are default value and disallow null. I was planning to add them after doing the work to set published_state based on current values of hidden/visible and is_stable. 

I would love to get clarity on if this is the direction we want to go because I have 4 other PRs built on top of it and I would like to avoid lots of doing and undoing if I can. Can you take a look before the end of day today?

Hamms commented on dashboard/app/models/script.rb:
```diff
@@ -121,6 +123,8 @@ def hide_pilot_scripts
       message: 'cannot start with a tilde or dot or contain slashes'
     }
 
+  validates :published_state, acceptance: {accept: SharedConstants::PUBLISHED_STATE.to_h.values.push(nil), message: 'must be nil, in_development, pilot, beta, preview or stable'}
```
I don't love that we're constructing this array inline, but I also don't have a specific reason why we shouldn't

dmcavoy replied:
The nil option will go away when we disallow null. Does that help with your concern?

Hamms replied:
it does!

Hamms reviewed (approved):
Holding off on default values until we've populated it with computed values makes sense to me, and all the other changes look good! One minor ruby nit, but feel free to merge without addressing if it's important.

davidsbailey reviewed (approved):
Thanks for your patience, Dani!

dmcavoy commented:
This is ready to be merged whenever staging is opened. The only failing test is the one that is the reason for staging being closed.
